runtime/compiler/optimizer/BoolArrayStoreTransformer.cpp

```diff
@@ -0,0 +1,485 @@
+ *
+ * bastore can be array store to either boolean or byte array. To
+ * figure out which bastore is for boolean array, the following steps are taken:
+ *
+ * 1. Normal IL generation pass solves the cases where the array base node is parm
+ *    ,field, or call. It also collects info about if there are stores of boolean
+ *    or bytes array type to any autos. If only one type of array or multi dimension
+ *    array ever appears in the IL trees, the bstorei nodes can only be this one type.
+ * 2. If arrays of both types exist in the IL trees, the transformer do a DFS traverse
+ *    of the CFG and tries to propagate auto types as best effort. This
+ *    approach might still leave some array's type unknown when there are stores of NULL
+ *    to an auto, or there are loops, or the base array node is from a multinewarray.
+ * 3. A runtime check is added for the array base node if the type is still unknown
+ *    for any bstorei nodes after the above steps. The check will very likely be folded
+ *    away by global value propagation and the overhead for important method body
+ *    should be minimum.
+ */
+
+#define OPT_DETAILS "O^O BOOL ARRAY STORE TRANSFORMER: "
+
+static char * getTypeName(TR_YesNoMaybe type, char * buffer)
+   {
+   switch (type)
+      {
+      case TR_yes:
+         strcpy(buffer, "[Z");
+         break;
+      case TR_no:
+         strcpy(buffer, "[B");
+         break;
+      case TR_maybe:
+         strcpy(buffer, "unknown type");
+         break;
+      }
+   return buffer;
+   }
+
+static void printTypeInfo(TR_BoolArrayStoreTransformer::TypeInfo *typeInfo, TR::Compilation *comp)
+   {
+   int localIndex = 0;
+   for (auto it = typeInfo->begin(); it != typeInfo->end(); it++)
+      {
+      if (*it != TR_maybe)
+         {
+         char buffer[15];
+         traceMsg(comp, "( local # %d: %s )  ", localIndex, getTypeName(*it, buffer));
+         }
+      localIndex++;
+      }
+   }
+
+TR_BoolArrayStoreTransformer::TR_BoolArrayStoreTransformer(NodeSet *bstoreiUnknownArrayTypeNodes, NodeSet *bstoreiBoolArrayTypeNodes)
+   :
+   _bstoreiUnknownArrayTypeNodes(bstoreiUnknownArrayTypeNodes),
+   _bstoreiBoolArrayTypeNodes(bstoreiBoolArrayTypeNodes),
+   _hasBoolArrayAuto(false),
+   _hasByteArrayAuto(false),
+   _numLocals(0)
+   {
+   _comp = TR::comp();
+   }
+
+void TR_BoolArrayStoreTransformer::perform()
+   {
+   if (comp()->getOption(TR_TraceILGen))
+      traceMsg(comp(), "<BoolArrayStoreTransformer>\n");
+
+   if (!_bstoreiUnknownArrayTypeNodes->empty() && _hasBoolArrayAuto)
+      {
+      if (_hasByteArrayAuto) // only need to iterate CFG if both byte and boolean array exist
+         findBoolArrayStoreNodes();
+      else // if only boolean array exist then all the bstorei nodes are operating on boolean array
+         {
+         _bstoreiBoolArrayTypeNodes->insert(_bstoreiUnknownArrayTypeNodes->begin(), _bstoreiUnknownArrayTypeNodes->end());
+         _bstoreiUnknownArrayTypeNodes->clear();
```

Comment:
I think this can erroneously truncate for a store into a `byte[]`:

    Object newByteArray(int n) { return new byte[n]; }
    ...
    void foo(int n, byte b) {
        boolean[] dummy = new boolean[n];
        ((byte[])newByteArray(n))[0] = b;
    }

Reply:
Added code to handle checkcast as well. thanks!

Reply:
It looks like this can happen again now that `isAnyDimension{Bool,Byte}ArrayNode()` ignore parameters' signatures:

    void foo(byte[] x, int i, byte b) {
        boolean[] dummy = new boolean[1];
        byte[] y = x;
        y[i] = b;
    }

Reply:
changed` isAnyDimension{Bool,Byte}ArrayNode()` to not use the default value.